[Merged by Bors] - Track hare message delay with prometheus histogram

piersy opened the pull request:
## Motivation

#4070

## Changes
Added 2 prometheus histograms per hare message type, one to track how late a message is and one to track how early a a message is relative to the time a round starts.

## Test Plan
We need to look at this in grafana with a more realistic network than the one generated by systest to see if the metrics will be useful.

I plan to follow up with an alternative approach that analyses the logs of message sending and receipt from multiple machines to see if the results correlate with what we get from the histograms.

The accuracy of the histogram approach will be correlated with the level of synchronisation of nodes' clocks, so if nodes tend to be quite out of sync then the histogram approach will not provide reliable data.

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

piersy committed:
Track hare message delay with prometheus histogram

piersy committed:
Merge branch 'develop' into piersy/network-propagation-time-histogram

countvonzero commented on hare/metrics.go:
```diff
@@ -45,4 +49,55 @@ var (
+		// Untracked message type
+		return
+	}
+	seconds := time.Since(clock.RoundEnd(t.round() - 1)).Seconds()
```
there is a difference btwn message type value and round number.

for example. status round (val 0) can have round number of 0/4/8/12. the param to `clock.RoundEnd()` expect the actual round number, which is recorded in `msg.Round()`

dshulyak replied:
i think the current implementation works correctly only for the first iteration (which is a bug)

piersy replied:
Are you referring to the round clock or my use of it here?

piersy replied:
Ok, thanks! Figuring out a fix.

countvonzero reviewed (commented):
will leave it to @dshulyak to review/approve.

dshulyak commented:
> The accuracy of the histogram approach will be correlated with the level of synchronisation of nodes' clocks, so if nodes tend to be quite out of sync then the histogram approach will not provide reliable data.

this is expected, but node clocks must remain in sync in order for protocol to work. there is a sub-protocol that checks that peers clock is in sync. i assume that most of the good nodes will be running ntp, small deviation is irrelevant and big outliers can be filtered out using statistical tools

dshulyak commented on hare/metrics.go:
```diff
@@ -45,4 +49,55 @@ var (
+	msgDelayTracker = newMessageDelayTracker()
 )
+
+func newMessageDelayTracker() *messageDelayTracker {
```
i think it will be easier to navigate one histogram with two labels (sign is optional for me though, we can understand if time is misconfigurated by looking at outliers). it is arguably also cleaner code, something like below

```go
var hareLatency = metrics.NewHistogramWithBuckets(
	"hare_latency_ms",
	subsystem,
	"Observed latency for hare message",
	[]string{"type", "sign"},
	prometheus.ExponentialBuckets(50, 2, 20),
)

func reportLatency(msgtype string, value time.Duration) {
	sign := "pos"
	if value < 0 {
		sign = "neg"
	}
	hareLatency.WithLabelValues(msgtype, sign).Observe(float64(value.Milliseconds()))
}
```

piersy replied:
Label sounds good. The reason I decided to track negative values was that from my understanding of [prometheus histograms](https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations) negative values are not supported, and if we were to simply discard them then we'd miss out on those observations.

piersy replied:
Actually, now understanding labels a bit better, it seems like making the message type a label would also be beneficial?

dshulyak commented on hare/metrics.go:
```diff
@@ -45,4 +49,55 @@ var (
+				namespace,
+				fmt.Sprintf("delta between expected time of receipt of %v message and actual receipt time", t),
+				[]string{},
+				prometheus.ExponentialBuckets(0.2, 2, 10),
```
i opted for ms for similar thing, do you think that working with floats is better?

piersy replied:
I was thinking of using ms too but then I read in the [prometheus label naming practices](https://prometheus.io/docs/practices/naming/) that you should use base units, which seemed a bit arbitrary to me but I switched :) I don't have any preference for float or non float, but in the end `Observe` expects a float, so I don't see any problem with using floats. Is there a problem?

piersy replied:
Happy to switch back to ms if that's going to be more inline with our current metrics

dshulyak replied:
i guess seconds might be somewhat easier to consume, with milliseconds it looks like this

![image](https://user-images.githubusercontent.com/1964576/225874466-0a1b61e2-a31f-4dbb-b790-dd9c6cae33ac.png)

dshulyak commented on hare/algorithm.go:
```diff
@@ -213,6 +213,9 @@ type consensusProcess struct {
 	once             sync.Once
+	// layerTime is used to calculate the difference between expected message
+	// arrival times and actual message arrival times
+	layerTime time.Time
```
expected hare start time can be always computed. 

hare_layer_time = genesis + layer*layer_duration + hare_wake_up

round_time = hare_layer_time + (round*iteration)*round_duration

layer and (round*iteration) is already encoded in the message, so time can be always computed without additional context

piersy replied:
Great!

piersy replied:
Actually, I wasn't even using it, since the round clock included everything that was required.

piersy committed:
Flip negative observations

My understanding is that observations for histograms should always be positive.
https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations

piersy committed:
Use labels in place of multipe metrics

Also use round instead of message type for calculating delays.

piersy committed:
Remove unused layer time field

piersy commented:
@dshulyak Updated, I think I've addressed all points

dshulyak commented on hare/metrics.go:
```diff
@@ -45,4 +47,32 @@ var (
+	if seconds < 0 {
+		sign = "neg"
+		// If the observation is negative make it positive.
+		seconds -= seconds
```
i think this will be always zero, as it translates to `seconds = seconds - seconds`

should probably be `seconds = -seconds`

piersy replied:
Haha, yes, well spotted!

dshulyak commented on hare/metrics.go:
```diff
@@ -45,4 +47,32 @@ var (
+		// If the observation is negative make it positive.
+		seconds -= seconds
+	}
+	hareLatency.WithLabelValues(t.String(), sign).Observe(seconds)
```
t.String() is in lowercase right?

piersy replied:
That should probably be lowercased. Thanks

dshulyak reviewed (approved):
please fix comment about neg

piersy committed:
Fix metric negation

piersy committed:
Ensure lowercase label names

piersy committed:
Merge branch 'develop' into piersy/network-propagation-time-histogram

piersy commented:
bors merge

countvonzero commented on hare/metrics.go:
```diff
@@ -45,4 +48,32 @@ var (
+		// If the observation is negative make it positive.
+		seconds = -seconds
+	}
+	hareLatency.WithLabelValues(strings.ToLower(t.String()), sign).Observe(seconds)
```
i think it would've been fine if you change the impl of t.String() to use lower case.

piersy replied:
I didn't change that since it would have affected a number of logs, but if its helpful then I can always change it in another PR.
